Web Inspector: REGRESSION(?): Sources Tab: async stack traces can sometimes be empty

dcrousso opened the pull request:
#### 312266846edf6e24d712c47bc872224f3d1bf785
<pre>
Web Inspector: REGRESSION(?): Sources Tab: async stack traces can sometimes be empty
<a href="https://bugs.webkit.org/show_bug.cgi?id=294065">https://bugs.webkit.org/show_bug.cgi?id=294065</a>

Reviewed by BJ Burg.

There&apos;s really no benefit in showing the developer that there was an (async) parent stack trace if nothing is in it (not counting the boundary).

* Source/JavaScriptCore/inspector/AsyncStackTrace.h:
* Source/JavaScriptCore/inspector/AsyncStackTrace.cpp:
(Inspector::AsyncStackTrace::buildInspectorObject const):
* Source/JavaScriptCore/inspector/ScriptCallStack.cpp:
(Inspector::ScriptCallStack::buildInspectorObject const):
Sanitize the generated `Inspector::Protocol::Console::StackTrace` object by skipping any that are empty (which might mean that there end up being no (async) parent stack traces to show).

* LayoutTests/inspector/console/message-stack-trace.html:
* LayoutTests/inspector/console/message-stack-trace-expected.txt:
* LayoutTests/inspector/console/resources/import-throw.js: Added.
* LayoutTests/inspector/console/resources/throw.js: Added.
Add a test for one reproducable scenario using dynamic `import(...)` as that seems to involve bunch of JS that&apos;s marked `ImplementationVisibility::Private` which results in (async) stack traces that only contain the `enqueueJob` boundary frame (which as stated above is pretty useless).

* LayoutTests/inspector/console/resources/log-console-stack-trace.js:
(logConsoleMessageStackTrace):
* LayoutTests/inspector/debugger/resources/async-stack-trace-test.js:
(logAsyncStackTrace):
* LayoutTests/inspector/debugger/resources/log-active-stack-trace.js:
(logStackTrace):
* LayoutTests/inspector/debugger/tail-deleted-frames/resources/stack-trace-utilities.js:
(logAsyncStackTrace):
Add more logging just in case this ever regresses in other situations.

Canonical link: <a href="https://commits.webkit.org/296069@main">https://commits.webkit.org/296069@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/527b58ba28e2edb64d736765f85bef7c0457148d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106366 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26115 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16513 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111567 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56962 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26782 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34618 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80793 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109370 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21217 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95990 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61121 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20711 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14089 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56402 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/99005 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90557 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14124 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114426 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/104983 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33504 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24701 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89864 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33868 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92219 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89568 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22996 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34452 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12269 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/29090 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33429 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38841 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/129295 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33175 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35222 "Found 1 new JSC stress test failure: wasm.yaml/wasm/gc/ref-i31-eq.js.default-wasm (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36528 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34773 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on LayoutTests/inspector/console/message-stack-trace-expected.txt:
```diff
@@ -57,3 +58,7 @@ ASYNC CALL STACK:
 2: [F] triggerUnhandledRejectionImplicitWithException
 
+-- Running test case: Console.StackTrace.UnhandledPromiseRejection.Module
+CALL STACK:
+0: [F] (anonymous function)
```
just to help clarify, without these changes this would look something like
```
CALL STACK:
0: [F] (anonymous function)
ASYNC CALL STACK:
0: --- enqueueJob ---
```

burg commented on LayoutTests/inspector/console/message-stack-trace.html:
```diff
@@ -41,6 +41,12 @@
+    let script = document.createElement("script");
+    script.type = "module";
+    script.src = "resources/import-throw.js";
+    document.body.appendChild(script);
```
🧙🏼‍♂️

burg commented on LayoutTests/inspector/console/resources/log-console-stack-trace.js:
```diff
@@ -22,16 +22,18 @@ TestPage.registerInitializer(() => {
             let truncated = stackTrace.truncated;
             stackTrace = stackTrace.parentStackTrace;
-            if (!callFrames || !callFrames.length)
+            if (!callFrames) {
```
As far as I can tell, none of the tests are hitting this code path. Is it needed? Or did some baselines not get added? Or is it defensive programming?

dcrousso replied:
yeah it's mainly just defensive

dcrousso committed:
Web Inspector: REGRESSION(?): Sources Tab: async stack traces can sometimes be empty
https://bugs.webkit.org/show_bug.cgi?id=294065

Reviewed by BJ Burg.

There's really no benefit in showing the developer that there was an (async) parent stack trace if nothing is in it (not counting the boundary).

* Source/JavaScriptCore/inspector/AsyncStackTrace.h:
* Source/JavaScriptCore/inspector/AsyncStackTrace.cpp:
(Inspector::AsyncStackTrace::buildInspectorObject const):
* Source/JavaScriptCore/inspector/ScriptCallStack.cpp:
(Inspector::ScriptCallStack::buildInspectorObject const):
Sanitize the generated `Inspector::Protocol::Console::StackTrace` object by skipping any that are empty (which might mean that there end up being no (async) parent stack traces to show).

* LayoutTests/inspector/console/message-stack-trace.html:
* LayoutTests/inspector/console/message-stack-trace-expected.txt:
* LayoutTests/inspector/console/resources/import-throw.js: Added.
* LayoutTests/inspector/console/resources/throw.js: Added.
Add a test for one reproducable scenario using dynamic `import(...)` as that seems to involve bunch of JS that's marked `ImplementationVisibility::Private` which results in (async) stack traces that only contain the `enqueueJob` boundary frame (which as stated above is pretty useless).

* LayoutTests/inspector/console/resources/log-console-stack-trace.js:
(logConsoleMessageStackTrace):
* LayoutTests/inspector/debugger/resources/async-stack-trace-test.js:
(logAsyncStackTrace):
* LayoutTests/inspector/debugger/resources/log-active-stack-trace.js:
(logStackTrace):
* LayoutTests/inspector/debugger/tail-deleted-frames/resources/stack-trace-utilities.js:
(logAsyncStackTrace):
Add more logging just in case this ever regresses in other situations.

Canonical link: https://commits.webkit.org/296069@main
